tezos/new_context/src/timings.rs

```diff
@@ -49,10 +70,621 @@ pub fn context_action(
+
+        if let Some(id) = Self::get_id_on_table(sql, table_name, &hash_string)? {
+            current.replace((id.to_string(), hash));
+            return Ok(());
+        };
+
+        let insert_query = format!(
+            "INSERT INTO {table} (hash) VALUES ('{hash}');",
+            table = table_name,
+            hash = hash_string
+        );
+
+        sql.execute(insert_query)?;
+
+        let id = Self::get_id_on_table(sql, table_name, &hash_string)?
+            .expect("Unable to find row after INSERT"); // This should never happen
+
+        current.replace((id.to_string(), hash));
+
+        Ok(())
+    }
+
+    fn get_id_on_table(
+        sql: &sqlite::Connection,
+        table_name: &str,
+        hash_string: &str,
+    ) -> Result<Option<i64>, SQLError> {
+        let query = format!(
+            "SELECT id FROM {table} WHERE hash = '{hash}';",
+            table = table_name,
+            hash = hash_string
+        );
+
+        let mut cursor = sql.prepare(query)?.into_cursor();
+
+        if let Some([Integer(id), ..]) = cursor.next()? {
+            return Ok(Some(*id));
+        };
+
+        Ok(None)
+    }
+
+    fn insert_checkout(
+        &mut self,
+        context_hash: ContextHash,
+        irmin_time: f64,
+        tezedge_time: f64,
+    ) -> Result<(), SQLError> {
+        self.checkouts.add(irmin_time, tezedge_time);
+        self.set_current_context(context_hash)?;
+        self.insert_action(&Action {
+            name: "checkout".to_string(),
+            key: vec![],
+            irmin_time,
+            tezedge_time,
+        })
+    }
+
+    fn insert_commit(&mut self, irmin_time: f64, tezedge_time: f64) -> Result<(), SQLError> {
+        self.commits.add(irmin_time, tezedge_time);
+        self.insert_action(&Action {
+            name: "commit".to_string(),
+            key: vec![],
+            irmin_time,
+            tezedge_time,
+        })?;
+        self.update_global_stats()
+    }
+
+    fn insert_action(&mut self, action: &Action) -> Result<(), SQLError> {
+        let block_id = self
+            .current_block
+            .as_ref()
+            .map(|(id, _)| id.as_str())
+            .unwrap_or("NULL");
+        let operation_id = self
+            .current_operation
+            .as_ref()
+            .map(|(id, _)| id.as_str())
+            .unwrap_or("NULL");
+        let context_id = self
+            .current_context
+            .as_ref()
+            .map(|(id, _)| id.as_str())
+            .unwrap_or("NULL");
+
+        let key = if action.key.is_empty() {
+            "NULL".to_string()
+        } else {
+            action.key.join("/")
+        };
+
+        let query = format!(
+            "
+        INSERT INTO actions
+          (name, key, irmin_time, tezedge_time, block_id, operation_id, context_id)
+        VALUES
+          ('{name}', '{key}', {irmin_time}, {tezedge_time}, {block_id}, {operation_id}, {context_id});
```

Comment:
Instead of buildings these with `format!`, we want to use Sqlite's placeholders, it is safer and faster. See the third snippet here for an example (the one with `?` placeholders): https://docs.rs/sqlite/0.26.0/sqlite/

Reply:
Btw the API for binding arguments in the `sqlite` crate looks like a PITA, `rusqlite` seems to provide a far more convenient way of doing it: https://docs.rs/rusqlite/0.25.1/rusqlite/